crates/search/src/provider.rs

```diff
@@ -0,0 +1,481 @@
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{
+    any::Any,
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use arrow_schema::{Field, Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    catalog::{Session, TableProvider},
+    common::{Column, DFSchema, JoinConstraint, JoinType},
+    datasource::{DefaultTableSource, TableType},
+    error::DataFusionError,
+    logical_expr::{
+        Filter, Join, LogicalPlan, Projection, Sort, SortExpr, TableProviderFilterPushDown,
+        TableScan,
+    },
+    physical_plan::ExecutionPlan,
+    prelude::Expr,
+    sql::TableReference,
+};
+
+use crate::{SEARCH_SCORE_COLUMN_NAME, index::SearchIndex};
+
+/// Performs a search on a given [`SearchIndex`] and combine with the underlying [`TableProvider`]
+/// if required by filters or additional columns in the projection.
+#[derive(Debug, Clone)]
+pub struct SearchQueryProvider {
+    pub search_index: Arc<dyn SearchIndex>,
+    pub table_provider: Arc<dyn TableProvider>,
+    pub query: String,
+    pub pre_limit: Option<usize>,
+}
+
+impl SearchQueryProvider {
+    pub fn new(
+        search_index: Arc<dyn SearchIndex>,
+        table_provider: Arc<dyn TableProvider>,
+        query: String,
+        limit: Option<usize>,
+    ) -> Self {
+        Self {
+            search_index,
+            table_provider,
+            query,
+            pre_limit: limit,
+        }
+    }
+
+    /// Check if the search index alone can satisfy the query (no join with base table needed)
+    fn search_index_table_is_sufficient(
+        &self,
+        search_index_table: &LogicalPlan,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+    ) -> Result<bool, DataFusionError> {
+        let search_index_columns: HashSet<String> = search_index_table
+            .schema()
+            .fields()
+            .iter()
+            .map(|f| f.name().to_string())
+            .collect();
+
+        // Check if projection can be satisfied
+        let source_schema = match projection {
+            None => self.schema(),
+            Some(indices) => {
+                let projected = self
+                    .schema()
+                    .project(indices)
+                    .map_err(|e| DataFusionError::ArrowError(e, None))?;
+                Arc::new(projected)
+            }
+        };
+
+        let columns_requested: HashSet<String> = source_schema
+            .fields()
+            .iter()
+            .map(|f| f.name().clone())
+            .collect();
+
+        let has_all_columns = search_index_columns.is_superset(&columns_requested);
```

Comment:
Should we return early w/o `calculating handleable_filters` if `has_all_columns` is false?

Reply:
Smart